examples/advanced/nlp-ner/jobs/bert_ncbi/app/custom/learners/bert_learner.py

```diff
@@ -0,0 +1,358 @@
+            a Shareable with the updated local model after running `execute()`
+        """
+        super().__init__()
+        self.aggregation_epochs = aggregation_epochs
+        self.train_task_name = train_task_name
+        self.num_labels = num_labels
+        self.lr = learning_rate
+        self.bs = batch_size
+        self.data_path = data_path
+        # client ID
+        self.client_id = None
+        # Epoch counter
+        self.epoch_of_start_time = 0
+        self.epoch_global = 0
+        # Training-related
+        self.train_loader = None
+        self.valid_loader = None
+        self.optimizer = None
+        self.device = None
+        self.model = None
+        self.writer = None
+        self.best_metric = 0.0
+        self.labels_to_ids = None
+        self.ids_to_labels = None
+
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        # when a run starts, this is where the actual settings get initialized for trainer
+        # set the paths according to fl_ctx
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+
+        # get and print the args
+        fl_args = fl_ctx.get_prop(FLContextKey.ARGS)
+        self.client_id = fl_ctx.get_identity_name()
+        self.log_info(
+            fl_ctx,
+            f"Client {self.client_id} initialized with args: \n {fl_args}",
+        )
+
+        # set local tensorboard writer for local validation score of global model
+        self.writer = SummaryWriter(app_dir)
+
+        # set the training-related contexts, this is task-specific
+        # get data from csv files
+        self.log_info(fl_ctx, f"Reading data from {self.data_path}")
+        df_train = pd.read_csv(os.path.join(self.data_path, self.client_id + "_train.csv"))
+        df_valid = pd.read_csv(os.path.join(self.data_path, self.client_id + "_val.csv"))
+        # get labels from data
+        labels = []
+        for x in df_train["labels"].values:
+            labels.extend(x.split(" "))
+        unique_labels = set(labels)
+        # check label length
+        if len(unique_labels) != self.num_labels:
+            self.system_panic(
+                f"num_labels {self.num_labels} need to align with dataset, actual data {len(unique_labels)}!", fl_ctx
+            )
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+        self.labels_to_ids = {k: v for v, k in enumerate(sorted(unique_labels))}
+        self.ids_to_labels = {v: k for v, k in enumerate(sorted(unique_labels))}
+        # set model
+        self.model = BertModel(num_labels=self.num_labels)
+        tokenizer = self.model.tokenizer
+        # set data
+        train_dataset = DataSequence(df_train, self.labels_to_ids, tokenizer=tokenizer)
+        valid_dataset = DataSequence(df_valid, self.labels_to_ids, tokenizer=tokenizer)
+        self.train_loader = DataLoader(train_dataset, num_workers=4, batch_size=self.bs, shuffle=True)
```

Comment:
num_workers seems a bit high for this task

Reply:
changed